### ISSUE TITLE
Prediction of racing car trajectory

### ISSUE THREAD

**Issue by the-zer07** [Unity-Technologies/ml-agents#2562]
Hi I'm trying to teach a car to run on the track, I'm using a frontal camera to take the visual observation and a "Lidar" sensors (360 raycast).
I have to teach the car a way to find the ideal trajectory but I don't know where to start.
I even thought if the edge detection from OpenCV or similar could be useful but I'm noob with Tensorflow and I don't know how to implement.
Please can someone help me?

**Comment by sterlingcrispin** [comment]
@the-zer07 what is your question? 

sounds like if you just reward it for forward motion on the track and punish it for going off the track it would learn to drive it

**Comment by the-zer07** [comment]
Yes that's it. Actually I punish him if he'll going off the track or collide against a barrier. Then I gain a little reward if he is still driving.
The car gain even a reward if the time of a sector or the entire track is new record.

**Comment by TheMartianLife** [comment]
Myself + the [Secret Lab](https://secretlab.institute/2019/07/11/installing-unity-ml-agents/) team just ran a tutorial teaching people new to Unity how to do a minimal working example of this at AIConf in San Jose. All the code and assets are up [on GitHub](https://github.com/parisba/ORM-AI-SJC-2019-Unity-ML-Agents) with [a full running sheet](https://github.com/parisba/ORM-AI-SJC-2019-Unity-ML-Agents/blob/master/Notes/RunningSheet.asciidoc) if you wanted something similar to look at?

Definitely an overfit problem--that agent would probably suck at any track that requires turning the other way--and the reward scheme could definitely be improved (say, by making completing a single lap the goal state) but the gist is there 🙂

**Comment by the-zer07** [comment]
> Myself + the [Secret Lab](https://secretlab.institute/2019/07/11/installing-unity-ml-agents/) team just ran a tutorial teaching people new to Unity how to do a minimal working example of this at AIConf in San Jose. All the code and assets are up [on GitHub](https://github.com/parisba/ORM-AI-SJC-2019-Unity-ML-Agents) with [a full running sheet](https://github.com/parisba/ORM-AI-SJC-2019-Unity-ML-Agents/blob/master/Notes/RunningSheet.asciidoc) if you wanted something similar to look at?
> 
> Definitely an overfit problem--that agent would probably suck at any track that requires turning the other way--and the reward scheme could definitely be improved (say, by making completing a single lap the goal state) but the gist is there 🙂

I think my problem is similar to your work. Now I pass the image by visual_observation in a function to get the edges and then the output as new visual_input.
But I still need to understand how to manage the throttle and brake.
I thought something like this:

    public override void AgentAction(float[] vectorAction, string textAction) {
        
        speed = carController.speed;

            if(isCollided) {
                AgentReset();
                Done();
            } else {
                steer = Mathf.Clamp(vectorAction[0], -1f, 1f);
                throttle = Mathf.Clamp(vectorAction[1], -1, 1f);
                
               carController.Move(steer, throttle);
                velocityReward = carController.speed / carController.maxspeed * 0.05f;
                AddReward(velocityReward);
            }
        }

In CarController I check:
    if(throttle >= 0)
        gas = throttle
    else
        brake = throttle

I don't know if this method is correct

**Comment by TheMartianLife** [comment]
Well the `CarController` I'm familiar with from the UnityStandardAssets the `Move()` function takes four parameters and looks like this

```csharp
Move(float steering, float accel, float footbrake, float handbrake)
```

So while our example--which prioritised being as simple as possible for Unity first-timers--employs a constant forward motion without braking and allows Agent control of steering only, you could just add VectorAction array positions so the Agent can control all four of these parameters at once (though it would be more prone to nonsense moves like accelerating and braking at once), else I would need to see your `CarController` convenience method that makes the final call to `Move()` to know for sure.

If your implementation is just ultimately applying `CarController.Move(steer, gas, brake, 0);` then it should be working just fine as is 👍

**Comment by the-zer07** [comment]
I don't use the UnityStandardAssets CarController, it's another package from the AssetsStore
